chalice/deploy/packager.py

```diff
@@ -208,3 +216,373 @@ def _add_chalice_lib_if_needed(self, project_dir, zip_fileobj):
+                valid_whls, missing_whls = self._download_dependencies(
+                    tempdir, requirements_file)
+                self._install_whls(tempdir, deps_dir, valid_whls)
+            if missing_whls:
+                raise MissingDependencyError(missing_whls)
+
+    def site_package_dir(self, project_dir):
+        # type: (str) -> str
+        """Return path to the site packages directory."""
+        deps_dir = self._osutils.joinpath(
+            project_dir, '.chalice', 'site-packages')
+        return deps_dir
+
+
+class Package(object):
+    PYPI_SDIST_EXTS = ['.zip', '.tar.gz']
+
+    def __init__(self, directory, filename):
+        # type: (str, str) -> None
+        self.dist_type = 'whl' if filename.endswith('whl') else 'sdist'
+        self.filename = filename
+        self._directory = directory
+        self._name, self._version = self._calculate_name_and_version()
+
+    @property
+    def data_dir(self):
+        # type: () -> str
+        # The directory format is {distribution}-{version}.data
+        return '%s-%s.data' % (self._name, self._version)
+
+    @property
+    def identifier(self):
+        # type: () -> str
+        return '%s==%s' % (self._name, self._version)
+
+    def __str__(self):
+        # type: () -> str
+        return '%s(%s)' % (self.identifier, self.dist_type)
+
+    def __repr__(self):
+        # type: () -> str
+        return str(self)
+
+    def __eq__(self, other):
+        # type: (Any) -> bool
+        return self.identifier == other.identifier
+
+    def __hash__(self):
+        # type: () -> int
+        return hash(self.identifier)
+
+    def _calculate_name_and_version(self):
+        # type: () -> Tuple[str, str]
+        if self.dist_type == 'whl':
+            # From the wheel spec (PEP 427)
+            # {distribution}-{version}(-{build tag})?-{python tag}-{abi tag}-
+            # {platform tag}.whl
+            name, version = self.filename.split('-')[:2]
+        else:
+            info_fetcher = SdistMetadataFetcher()
+            name, version = info_fetcher.get_package_name_and_version(
+                self._directory, self.filename)
+        return name, version
+
+
+class SdistMetadataFetcher(object):
+    """This is the "correct" way to get name and version from an sdist."""
+    # https://git.io/vQkwV
+    _SETUPTOOLS_SHIM = (
+        "import setuptools, tokenize;__file__=%r;"
+        "f=getattr(tokenize, 'open', open)(__file__);"
+        "code=f.read().replace('\\r\\n', '\\n');"
+        "f.close();"
+        "exec(compile(code, __file__, 'exec'))"
+    )
+
+    def __init__(self, osutils=None):
+        # type: (OSUtils) -> None
+        if osutils is None:
+            osutils = OSUtils()
+        self._osutils = osutils
+
+    def _parse_pkg_info_file(self, filepath):
+        # type: (str) -> Message
+        with open(filepath, 'r') as f:
+            data = f.read()
+        parser = FeedParser()
```

Comment:
FWIW that's the API that the stdlib provides for parsing a PKG-INFO file, largely for historical reasons. It could use the pkg-info library though.